pkg/telemetry/exporter.go

```diff
@@ -0,0 +1,104 @@
+// Copyright © 2023 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telemetry
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+func exporterFromConfig(ctx context.Context, config *Config) (sdktrace.SpanExporter, error) {
+	switch strings.ToLower(config.Exporter) {
+	case "otlp":
+		return initOTLPExporter(ctx, &config.CollectorConfig)
+
+	case "writer":
+		return initWriterExporter(&config.WriterConfig)
+
+	default:
+		return nil, errors.New("unknown telemetry exporter")
+	}
+}
+
+func initOTLPExporter(ctx context.Context, config *CollectorConfig) (sdktrace.SpanExporter, error) {
+	var opts []grpc.DialOption
+	switch {
+	case config.Insecure:
+		opts = append(opts, grpc.WithTransportCredentials(insecure.NewCredentials()))
+
+	default:
+		tlsConfig := http.DefaultTransport.(*http.Transport).TLSClientConfig
+		switch {
+		case tlsConfig == nil:
+			tlsConfig = new(tls.Config)
+			fallthrough
+
+		case tlsConfig.RootCAs == nil:
+			var err error
+			if tlsConfig.RootCAs, err = x509.SystemCertPool(); err != nil {
+				tlsConfig.RootCAs = x509.NewCertPool()
+			}
+		}
+
+		if err := config.TLS.ApplyTo(tlsConfig); err != nil {
+			return nil, err
+		}
+		opts = append(opts, grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)))
+	}
+
+	conn, err := grpc.DialContext(ctx, config.EndpointURL, append(opts, grpc.WithBlock())...)
```

Comment:
Add `grpc.FailOnNonTempDialError(true)` as dial option here

Reply:
By default, this value is `localhost:<something>`; so does this mean TTS relies on this endpoint to be there?

Reply:
OK, but how can I see from `DefaultTelemetryConfig` that it isn't enabled?

I understand about the Redis reference, but Redis is required for TTS.

Reply:
Boolean `Enable` is false by default, but it could be set explicitly to false in the default struct.